### PR TITLE
Fix mDNS discovery (bug #577 #1745)

### PR DESCRIPTION
**Pull request by lolodomo**
One JmDNS instance is now created for each local IP address.
Loopback and Link-local addresses are ignored.
Only network interfaces that are up at openHAB startup are considered.

This change is important when running openHAB on a machine having one IP
v4 and one IP v6. On such a machine, two things will now work correctly:
- the discovery of IP v4 services
- the publishing of openHAB services on IP v4 and IP v6

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=577

**Comment by lolodomo** [comment]
Ok, what's wrong with my account ?!
You can check that CLA is ok for lg.hc@free.fr
Is signed-off-by case sensitive ?
What else ?

**Comment by lolodomo** [comment]
Even with Signed-Off-By, it does not work.
What a waste of time !

**Comment by kaikreuzer** [comment]
Don't worry! The automatic CLA check sometimes fails, but the committers can always manually check the contributors CLA and merge PRs nonetheless.
I can confirm that there is a valid CLA for you.

**Comment by lolodomo** [comment]
Fix for issue #577

**Comment by maggu2810** [comment]
I don't know If the "Signed-off-by" is compared case-sensitive, but I think it should work if you add the angle brackets around the mail address (it has been the case by another person some time ago).

If you are using "-s" on commit the signed-off-by header should be created correctly by git.

**Comment by lolodomo** [comment]
I added in my PR the fix for #1745.
And you're right, @maggu2810 , with angle brackets around the mail address, the IP validation check seems to be ok now.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java [hunk 1]
```diff
@@ -68,59 +73,41 @@ public void run() {
...
             }
         }, 0, TimeUnit.SECONDS);
-        for (MDNSDiscoveryParticipant participant : participants) {
-            mdnsClient.getClient().addServiceListener(participant.getServiceType(), this);
-        }
-    }
-
-    @Override
-    protected void stopBackgroundDiscovery() {
-        for (MDNSDiscoveryParticipant participant : participants) {
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java]: why did you remove this method?

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java [hunk 2]
```diff
@@ -24,25 +31,60 @@
...
+                addresses.add(address);
+            }
+        }
+        return addresses;
+    }
 
     @Override
-    public JmDNS getClient() {
-        return jmdns;
+    public Set<JmDNS> getClientInstances() {
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java]: Did you check whether this API change breaks any existing bindings in openHAB 2?
If it does NOT break any, it means that getClient() is only used by the MDNSDiscoveryService implementation. If this is indeed the case, I wonder if there is at all a need to expose the full set of JmDNS instances. Maybe we could simply return some simple wrapper that mimicks a JmDNS instance, but forwards the calls (e.g. to addServiceListener) to all internal instances. This would make the code for the caller much simpler, as there wouldn't be a need for loops anymore.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java (hunk 1 above)
Comment (lolodomo) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java]: Because there was no more code inside !
I can restore it, empty, if you like.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java (hunk 2 above)
Comment (lolodomo) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java]: Yes, I hesiated to do that. If you think it is preferable, I will change that.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java (hunk 1 above)
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java]: Ok, then let me ask you differently: Why did you remove the code within this method? If no background discovery is wished, no listeners should be registered and no devices should be found.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java (hunk 2 above)
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java]: I am not sure whether it is a valid approach or not, but it would look simpler for users. I don't think that we will have use cases where callers would need to have individual access to the different network interfaces. And if, we can maybe then add a dedicated additional method for it.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java (hunk 1 above)
Comment (lolodomo) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java]: I explained why in issue #1745.
One listener is already registered in setMDNSClient (when binding the service) and always active. Managing a second listener has no interest, methods are called twice during the discovery.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java (hunk 2 above)
Comment (lolodomo) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java]: Ok for that, I will implement the change.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java (hunk 1 above)
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java]: ok, thanks for the pointer - I have commented on #1745.

**Comment by lolodomo** [comment]
Ok, PR updated to take account the different remarks.

**Comment by lolodomo** [comment]
With my last PR update and assuming the bug in JmDNS will be fixed one day, that should be working perfectly.

Until JmDNS is fixed, when background discovery is enabled, listener will be registered twice but hopefully OH inbox is enough clever to not duplicate things. Another consequence is that if background discovery is disabled after OH is started, one listener will remain registered. As a workarround I added a control to avoid creating new discovery things when the background discovery is not enabled.

**Comment by lolodomo** [comment]
Don't merge this PR, I saw an exception when openHAB is shutdown. Will try to understand what's wrong.

**Comment by kaikreuzer** [comment]
@lolodomo What is the status of this PR?

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java [hunk 3]
```diff
@@ -24,29 +34,164 @@
...
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void registerService(ServiceDescription description) throws IOException {
+        for (JmDNS instance : jmdnsInstances) {
+            logger.debug("Registering new service {} at port {} (instance {})", description.serviceType,
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java]: I'd suggest to change this to
```
            logger.debug("Registering new service {} at {}:{} ({})", description.serviceType,
                    instance.getInetAddress().getHostAddress(), description.servicePort, instance.getName());
```

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java [hunk 4]
```diff
@@ -24,29 +34,164 @@
...
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void unregisterService(ServiceDescription description) {
+        for (JmDNS instance : jmdnsInstances) {
+            logger.debug("Unregistering service {} at port {} (instance {})", description.serviceType,
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSClientImpl.java]: I'd suggest to change this to
```
            logger.debug("Unregistering service {} at {}:{} ({})", description.serviceType,
                    instance.getInetAddress().getHostAddress(), description.servicePort, instance.getName());
```

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSServiceImpl.java [hunk 5]
```diff
@@ -44,13 +42,17 @@ public void setMDNSClient(MDNSClient client) {
...
                 @Override
                 public void run() {
-                    for (ServiceInfo serviceInfo : servicesToRegisterQueue) {
+                    logger.debug("Registering {} queued services", servicesToRegisterQueue.size());
+                    for (ServiceDescription description : servicesToRegisterQueue) {
                         try {
-                            logger.debug("Registering new service " + serviceInfo.getType() + " at port "
-                                    + String.valueOf(serviceInfo.getPort()));
-                            mdnsClient.getClient().registerService(serviceInfo);
+                            logger.debug("Registering new service {} at port {}", description.serviceType,
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSServiceImpl.java]: you can remove this log message as it duplicates the ones in mdnsClient.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSServiceImpl.java [hunk 6]
```diff
@@ -72,23 +74,20 @@ public void registerService(final ServiceDescription description) {
...
             Runnable runnable = new Runnable() {
                 @Override
                 public void run() {
-                    ServiceInfo serviceInfo = ServiceInfo.create(description.serviceType, description.serviceName,
-                            description.servicePort, 0, 0, description.serviceProperties);
                     try {
-                        logger.debug("Registering new service " + description.serviceType + " at port "
-                                + String.valueOf(description.servicePort));
-                        mdnsClient.getClient().registerService(serviceInfo);
+                        logger.debug("Registering new service {} at port {}", description.serviceType,
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSServiceImpl.java]: you can remove this log message as it duplicates the ones in mdnsClient.

File: bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSServiceImpl.java [hunk 7]
```diff
@@ -101,22 +100,18 @@ public void run() {
      */
     @Override
     public void unregisterService(ServiceDescription description) {
-        if (mdnsClient == null) {
-            return;
+        if (mdnsClient != null) {
+            logger.debug("Unregistering service {} at port {}", description.serviceType, description.servicePort);
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/internal/MDNSServiceImpl.java]: you can remove this log message as it duplicates the ones in mdnsClient.

**Comment by lolodomo** [comment]
I will take into account your comments relative to debug logs.
This PR was OK until I discovered that an exception can occur when shutting down openHAB. I will have to investigate. Not sure that it is due to my changes.

**Comment by lolodomo** [comment]
As you created a new issue #1932, I propose to exclude from the perimeter of this PR the fix of this new issue.

**Comment by kaikreuzer** [comment]
fine for me - so you say this PR is ok to be merged then?

**Comment by lolodomo** [comment]
I first have to consider your comments from today.
I will update my PR before the end of the day.

**Comment by lolodomo** [comment]
PR updated, ready for merging.

**Comment by kaikreuzer** [comment]
Thanks!

> assuming the bug in JmDNS will be fixed one day

Do we have an issue at JmDNS for it? If not, could you enter one, so that this isn't forgotten?

**Comment by kaikreuzer** [comment]
Hm, I should have had a closer look at JmDNS much earlier, I guess.

Wouldn't the right approach to support multiple interfaces be to use JmmDNS instead of JmDNS? This is what the JavaDoc of JmmDNS says:

>  Java Multihomed Multicast DNS
>
> Uses an underlying {@link JmDNS} instance for each {@link InetAddress} found on this computer.
> This class will monitor network topology changes, and will create or destroy JmDNS instances as required.

Shall we envisage to move to this implementation?

**Comment by maggu2810** [comment]
At least the JavaDoc sounds as it would be the correct choose.
